crates/trie/trie/src/proof_v2/mod.rs

```diff
@@ -73,18 +373,72 @@ where
         #[cfg(not(debug_assertions))]
         let targets = targets.into_iter();
 
+        // Ensure initial state is cleared. By the end of the method call these should be empty once
+        // again.
+        debug_assert!(self.branch_stack.is_empty());
+        debug_assert!(self.branch_path.is_empty());
+        debug_assert!(self.child_stack.is_empty());
+
         // Silence unused variable warning for now
         let _ = targets;
 
-        unimplemented!()
+        let mut proof_nodes = Vec::new();
+        let mut hashed_cursor_current = self.hashed_cursor.seek(B256::ZERO)?;
+        loop {
+            trace!(target: TRACE_TARGET, ?hashed_cursor_current, "proof_inner loop");
+
+            // Fetch the next leaf from the hashed cursor, converting the key to Nibbles and
+            // immediately creating the DeferredValueEncoder so that encoding of the leaf value can
+            // begin ASAP.
+            let Some((key, val)) = hashed_cursor_current.map(|(key_b256, val)| {
+                debug_assert_eq!(key_b256.len(), 32);
+                // SAFETY: key is a B256 and so is exactly 32-bytes.
+                let key = unsafe { Nibbles::unpack_unchecked(key_b256.as_slice()) };
+                let val = value_encoder.deferred_encoder(key_b256, val);
+                (key, val)
+            }) else {
+                break
+            };
+
+            self.add_leaf(key, val)?;
+            hashed_cursor_current = self.hashed_cursor.next()?;
+        }
+
+        // Once there's no more leafs we can pop the remaining branches, if any.
+        while !self.branch_stack.is_empty() {
+            self.pop_branch()?;
+        }
+
+        // At this point the branch stack should be empty. If the child stack is empty it means no
+        // keys were ever iterated from the hashed cursor in the first place. Otherwise there should
+        // only be a single node left: the root node.
+        debug_assert!(self.branch_stack.is_empty());
+        debug_assert!(self.branch_path.is_empty());
+        debug_assert!(self.child_stack.len() < 2);
+
+        // Determine the root node based on the child stack, and push the proof of the root node
+        // onto the result stack.
+        let root_node = if let Some(node) = self.child_stack.pop() {
+            node.into_trie_node(&mut self.rlp_encode_buf)?
+        } else {
+            TrieNode::EmptyRoot
+        };
+
+        proof_nodes.push(ProofTrieNode {
+            path: Nibbles::new(), // root path
+            node: root_node,
+            masks: TrieMasks::none(),
+        });
```

Comment:
This will be removed in the next PR when real proof retention based on `targets` is added, but it's useful for now to verify the implementation